enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate conditions to OLM that it cannot infer otherwise.
+
+#### Story 2
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade or uninstall my operator until I tell it otherwise.
+
+## Design Details
+
+Operators will communicate with OLM by setting [conditions](https://dev.to/maelvls/what-the-heck-are-kubernetes-conditions-for-4je7) on an owned CustomResource. Conditions are used by many Kubernetes Resources to provide users with additional insight into the state of the resource. For those unfamiliar with conditions, they typically follow the following format:
+
+| Field | Description |
+| ----- | ----------- |
+| Type |  A one-word description of the condition. |
+| Status | Status is the status of the condition. Can be True, False, Unknown. |
+| Reason | Unique, one-word, CamelCase reason for the condition's last transition. |
+| Message | Human-readable message indicating details about last transition. Typically used by cluster admins for debugging purposes. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
+
+### OLM Supported Conditions
+
+Although an operator may set any condition on a CustomResource that it owns, there will be a subset of "OLM Supported Conditions" which will influence the behavior of OLM when present. The list of "OLM Supported Conditions" can be seen in the table below:
+
+| Condition Name | Effect |
+| -------------- | ------ |
+| NotInterruptible | When true, OLM will not upgrade the operator. The pod associated with the operator may still be killed by external forces.|
+| Important | OLM will highlight this condition to the Cluster Admin.|
+
+### Why Communicate through CustomResources
+
+There are many reasons why operators should communicate their conditions through owned CustomResources:
+
+- CustomResources are the main entrypoint to an operator and operators are already expected to update the status of the CustomResource to communicate "Actual State".
+- By writing the condition to the status of an owned CustomResource, it becomes very clear which CR is responsible for changing OLM behavior.
+- Introducing an OLM owned CRD forces operators to integrate with OLM at a much deeper level than writing to a resource that they already manage. If OLM were to provide a CRD, the operators would be required to detect if it is being managed by OLM and begin writing to the API. By writing to an owned CR, the Operator does not require OLM specific code and can be run independently of OLM.
+  - As a bonus, OLM will not increase the scope of APIs that it owns.
+
+### How Operators will opt-in
+
+Operators must opt-in to "OLM Supported Conditions" by adding an annotation to their owned CRD, where the key is the defined by OLM and the value is the `type` of the condition on the CR.
```

Comment:
What if the CR that has the not condition is from a required CRD, and not an owned CRD? 

For example, vault has a requirement on etcd. Say the vault migration fails due to a etcd cluster in a bad state - the vault operator would write the condition to the vault CR? That seems to work. I feel like there is some additional complexity in this area that is worth looking at - are we covering all the cases.

Reply:
This is a great point and should be discussed in greater detail in this enhancement. I believe that OLM should consider if an operator can be upgraded when calculating a new `Generation`.